### PR TITLE
Use one package for different operating systems

### PR DESCRIPTION
**Pull request by pustovalov**
refers to https://github.com/loeffel-io/ls-lint/issues/8

saw this approach in this package: https://github.com/Arkweid/lefthook

Demo package: https://yarnpkg.com/package/pustovalov-ls-lint

Windows log: https://gist.github.com/pustovalov/6ff9d33980e546d04b5ae8c860fb0a74#file-win32-log

macOS log: https://gist.github.com/pustovalov/6ff9d33980e546d04b5ae8c860fb0a74#file-darwin-log

### CODE REVIEW

**Comment by loeffel-io** [comment]
@pustovalov First of all: thank you very much, you are the man! 🚀 

I already read about this a lot yesterday night! 
fyi: I checked out https://github.com/electron/electron and https://github.com/evanw/esbuild. 

Let's make this happen today! 

Lets go: (some thoughts and facts, intro to my code review of your PR)

- First of all i do not like the idea of spawn binaries with node, but i think there is no way around. In this case i would love to still **also** support macos, linux (and windows) npm packages **if there is a major performance difference**
- Sadly the name `ls-lint` is already reserved. Maybe we should name the package `ls-lint-os` or something similar. Maybe you have a good idea for this?

File: Makefile [hunk 1]
```diff
@@ -13,15 +13,9 @@ build:
 	GOOS=windows GOARCH=amd64 go build -o ls-lint-windows.exe
 
 build-npm:
```
Comment (loeffel-io) [inline Makefile]: referending to my [comment](https://github.com/loeffel-io/ls-lint/pull/11#issuecomment-598664338) i would love to have separated make commands for every platform. 

should look like:

```
build-npm:
	make build-npm-darwin
	make build-npm-linux
	make build-npm-windows

build-npm-darwin:
	cp ls-lint-darwin npm/bin/ls-lint-darwin
	chmod +x npm/bin/ls-lint-darwin

build-npm-linux:
	cp ls-lint-linux npm/bin/ls-lint-linux
	chmod +x npm/bin/ls-lint-linux

build-npm-windows:
	cp ls-lint-windows.exe npm/bin/ls-lint-windows.exe
	chmod +x npm/bin/ls-lint-windows.exe
```

**Comment by pustovalov** [comment]
What about `@loeffel/ls-lint`?

File: npm/ls-lint/package.json [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  "os": [
+    "darwin",
+    "linux",
+    "win32"
+  ],
+  "cpu": [
+    "x64"
+  ],
+  "scripts": {
+    "postinstall": "node postinstall.js"
```
Comment (loeffel-io) [inline npm/ls-lint/package.json]: would love to rename: 

`postinstall.js` to `install.js`
`index.js` to `cli.js`

File: npm/ls-lint/package.json [hunk 3]
```diff
@@ -0,0 +1,22 @@
+{
+  "name": "ls-lint",
```
Comment (loeffel-io) [inline npm/ls-lint/package.json]: referending to my comment we have to change the name here

File: README.md [hunk 4]
```diff
@@ -90,12 +90,10 @@ ls-lint-windows.exe
 
 ```bash
 # global
-npm install -g ls-lint-linux # for Linux
```
Comment (loeffel-io) [inline README.md]: referending to my comment we have to change the name here

**Comment by loeffel-io** [comment]
@pustovalov i just created a `ls-lint` organisation.  

`@loeffel/ls-lint` is a great idea, but i would prefer something without my name (loeffel is in english a spoon: https://images-na.ssl-images-amazon.com/images/I/61Y-85Ei4-L._AC_SL1500_.jpg 😂) 

What do you think about `@ls-lint/ls-lint`?

File: npm/ls-lint/postinstall.js [hunk 5]
```diff
@@ -0,0 +1,24 @@
+const { spawn } = require("child_process");
+const { join } = require("path");
+
+const isCI = process.env.CI;
```
Comment (loeffel-io) [inline npm/ls-lint/postinstall.js]: i think we do not need the CI ENV variable

File: npm/ls-lint/postinstall.js [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+if (!isCI) {
+  if (process.platform === 'darwin') {
```
Comment (loeffel-io) [inline npm/ls-lint/postinstall.js]: please use switch case here, cause there could be some multi cases like here:
 
https://github.com/electron/electron/blob/master/npm/install.js#L67

File: npm/ls-lint/bin/index.js [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+  );
+
+  child.on('close', function(code) {
+    if (code !== 0) {
+      process.exit(1);
+    }
+  });
+}
+
+if (process.platform === 'darwin') {
```
Comment (loeffel-io) [inline npm/ls-lint/bin/index.js]: please use switch case here, cause there could be some multi cases like here:

https://github.com/electron/electron/blob/master/npm/install.js#L67

File: npm/ls-lint/postinstall.js [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+    binary = 'ls-lint-darwin';
+  } else if (process.platform === 'linux') {
+    binary = 'ls-lint-linux';
+  } else if (process.platform === 'win32') {
+    binary = 'ls-lint-windows.exe';
+  } else {
+    console.log("Unsupported OS");
```
Comment (loeffel-io) [inline npm/ls-lint/postinstall.js]: we should use a error message like: 

console.log('ls-lint builds are not available on platform: ' + process.platform)

File: npm/ls-lint/bin/index.js [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+  spawnCommand('ls-lint-darwin');
+} else if (process.platform === 'linux') {
+  spawnCommand('ls-lint-linux');
+} else if (process.platform === 'win32') {
+  spawnCommand('ls-lint-windows.exe');
+} else {
+  console.log("Unsupported OS");
```
Comment (loeffel-io) [inline npm/ls-lint/bin/index.js]: we should use a error message like: 

console.log('ls-lint builds are not available on platform: ' + process.platform)

**Comment by loeffel-io** [comment]
We also have to change the job in the `.drone.yml` file to:

```yaml:
  - name: npm
    image: plugins/npm
    settings:
      username: loeffel-io
      password:
        from_secret: NPM_PASSWORD
      email: lucas@loeffel.io
      folder: npm/ls-lint
    when:
      event: tag
```

**Comment by loeffel-io** [comment]
close #8 
close #1

**Comment by loeffel-io** [comment]
I tested the demo package from you with this result: 

```bash
➜  npm-package npm add pustovalov-ls-lint

> pustovalov-ls-lint@0.0.4 postinstall /Users/lucas.loeffel/ls-lint-tests/node_modules/pustovalov-ls-lint
> node postinstall.js

**2020/03/13 12:29:14 open .ls-lint.yml: no such file or directory**
npm WARN saveError ENOENT: no such file or directory, open '/Users/lucas.loeffel/ls-lint-tests/package.json'
npm WARN enoent ENOENT: no such file or directory, open '/Users/lucas.loeffel/ls-lint-tests/package.json'
npm WARN ls-lint-tests No description
npm WARN ls-lint-tests No repository field.
npm WARN ls-lint-tests No README data
npm WARN ls-lint-tests No license field.

+ pustovalov-ls-lint@0.0.4
updated 1 package and audited 1 package in 0.528s
found 0 vulnerabilities
```

I think we just do not need the postinstall command right? There is no action required after the installation in my opinion

**Comment by loeffel-io** [comment]
Tested this on macos and a digitalocean ubuntu droplet: works perfectly! 🚀

**Comment by pustovalov** [comment]
New demo package: `pustovalov-ls-lint@0.0.8`

File: npm/package.json [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+  "main": "./bin/cli.js",
+  "bin": {
+    "ls-lint": "./bin/cli.js"
+  },
+  "os": [
+    "darwin",
+    "linux",
+    "win32"
+  ],
+  "cpu": [
```
Comment (loeffel-io) [inline npm/package.json]: i think we should remove this one

Comment (loeffel-io) [reply npm/package.json]: i read about that, i think it makes sense! sry about that

**Comment by loeffel-io** [comment]
Thank you so much! I will check the demo package v0.0.8 now on all platforms 👍

**Comment by loeffel-io** [comment]
Do you have access to a windows machine currently?

**Comment by pustovalov** [comment]
```
PS C:\check> yarn add pustovalov-ls-lint
yarn add v1.19.1
[1/4] Resolving packages...
[2/4] Fetching packages...
info fsevents@1.2.4: The platform "win32" is incompatible with this module.
info "fsevents@1.2.4" is an optional dependency and failed compatibility check. Excluding it from installation.
[3/4] Linking dependencies...
[4/4] Building fresh packages...

success Saved lockfile.
success Saved 1 new dependency.
info Direct dependencies
└─ pustovalov-ls-lint@0.0.8
info All dependencies
└─ pustovalov-ls-lint@0.0.8
Done in 10.28s.
PS C:\check> yarn ls-lint
yarn run v1.19.1
$ C:\check\node_modules\.bin\ls-lint
2020/03/13 14:11:54 .\benchmarks\dbmon\ENV.js failed for rules: kebabcase
2020/03/13 14:11:54 .\benchmarks\ssr\renderToStream.js failed for rules: kebabcase
2020/03/13 14:11:54 .\benchmarks\ssr\renderToString.js failed for rules: kebabcase
2020/03/13 14:11:54 .\src\compiler\create_compiler.js failed for rules: kebabcase
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
PS C:\check>
```

**Comment by loeffel-io** [comment]
Perfect! 

Tested this also on ubuntu and macos. I trust you with the windows version! 

Lets publish this npm package!

**Comment by loeffel-io** [comment]
DONE: https://www.npmjs.com/package/@ls-lint/ls-lint ❤️ 
Thank you very much @pustovalov

**Comment by imlinus** [comment]
Absolute legends! 💯

**Comment by pustovalov** [comment]
ls-lint 1.5.0 log:
https://gist.github.com/pustovalov/aaccb1dea2c8ab762e826d522324d14f

**Pull request merged**